flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategyTest.java

```diff
@@ -0,0 +1,194 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.strategy;
+
+import org.apache.flink.api.common.InputDependencyConstraint;
+import org.apache.flink.runtime.execution.ExecutionState;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionType;
+import org.apache.flink.runtime.jobgraph.DistributionPattern;
+import org.apache.flink.runtime.jobgraph.IntermediateDataSet;
+import org.apache.flink.runtime.jobgraph.IntermediateResultPartitionID;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.jobgraph.JobVertex;
+import org.apache.flink.runtime.scheduler.ExecutionVertexDeploymentOption;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Unit test for {@link LazyFromSourcesSchedulingStrategy}.
+ */
+public class LazyFromSourcesSchedulingStrategyTest extends TestLogger {
+
+	/**
+	 * Tests that when start scheduling lazy from sources scheduling strategy will start input vertices in scheduling topology.
+	 */
+	@Test
+	public void testStartScheduling() {
+		JobVertex jobVertex1 = new JobVertex("vertex#1");
+		JobVertex jobVertex2 = new JobVertex("vertex#2");
+		JobGraph graph = new JobGraph(jobVertex1, jobVertex2);
+		jobVertex2.connectNewDataSetAsInput(jobVertex1, DistributionPattern.ALL_TO_ALL, ResultPartitionType.BLOCKING);
+
+		TestingSchedulingTopology schedulingTopology = new TestingSchedulingTopology();
+
+		for (int i = 0; i < 3; i++) {
+			schedulingTopology.addSchedulingVertex(new TestingSchedulingVertex(jobVertex1.getID(), i));
+			schedulingTopology.addSchedulingVertex(new TestingSchedulingVertex(jobVertex2.getID(), i));
+		}
+
+		TestingSchedulerOperation testingSchedulerOperation = new TestingSchedulerOperation();
+		LazyFromSourcesSchedulingStrategy schedulingStrategy = new LazyFromSourcesSchedulingStrategy(
+			testingSchedulerOperation,
+			schedulingTopology,
+			graph);
+
+		schedulingStrategy.startScheduling();
+
+		assertEquals(3, testingSchedulerOperation.getScheduledVertices().size());
+	}
+
+	/**
+	 * Tests that when on execution state change will start available downstream vertices.
+	 * vertex#0    vertex#1
+	 *       \     /
+	 *        \   /
+	 *         \ /
+	 *  (BLOCKING, ALL)
+	 *     vertex#3      vertex#2
+	 *           \        /
+	 *            \      /
+	 *             \    /
+	 *          (BLOCKING, ANY)
+	 *             vertex#4
+	 *                |
+	 *                |
+	 *                |
+	 *            (PIPELINED)
+	 *             vertex#5
+	 */
+	@Test
+	public void testOnExecutionStateChange() {
+		TestingSchedulingTopology schedulingTopology = new TestingSchedulingTopology();
```

Comment:
I think this method is a bit too long. It could be split into more units (see Clean Code by Robert C Martin, Chapter 3).

Reply:
OK